File: LEET CODE QUESTIONS/Spiral Matrix.cpp
//54. Spiral Matrix
//Medium
//9.9K
//972
//Companies
//Given an m x n matrix, return all elements of the matrix in spiral order.


//Example 1:


//Input: matrix = [[1,2,3],[4,5,6],[7,8,9]]
//Output: [1,2,3,6,9,8,7,4,5]
//Example 2:

//Input: matrix = [[1,2,3,4],[5,6,7,8],[9,10,11,12]]
//Output: [1,2,3,4,8,12,11,10,9,5,6,7]

class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) 
    {
       int m=matrix.size();
         vector<int>res;
          if(m==0)
          return res;
        int n=matrix[0].size();
        
        int rowstart=0;
        int rowend=m;
        int colstart=0;
        int colend=n;
        
        int row, col;
        while(rowstart<rowend && colstart<colend)
        {
          
            for(col=colstart;col<colend;col++)
            res.push_back(matrix[rowstart][col]);
            rowstart++;
            
            for(row=rowstart;row<rowend;row++)
            res.push_back(matrix[row][colend-1]);
            colend--;
            
            if(rowstart<rowend)
            {
                for(col=colend-1;col>=colstart;col--)
                res.push_back(matrix[rowend-1][col]);
                rowend--;
            }
            
            if(colstart<colend)
            {
                for(row=rowend-1;row>=rowstart;row--)
                res.push_back(matrix[row][colstart]);
                colstart++;
            }
            
        }
        return res;       
    }
};

File: JAVA/Spiral Matrix.java
import java.util.*;

public class Main {
	public static void main(String[] args) {
	    Scanner sc = new Scanner(System.in);
	    int n = sc.nextInt();
	    int m = sc.nextInt();
	    
	    int [][] matrix = new int[n][m];
	    
	    // input
	    for(int i=0;i<n;i++){
	        for(int j=0;j<m;j++){
	            matrix[i][j] = sc.nextInt();
	        }
	    }
	    
	    System.out.println("The spiral order matrix is : ");
	    
	    int row_start = 0, row_end = n-1;
	    int col_start = 0, col_end = m-1;
	    
	    // To print spiral order matrix
	    
	    while(row_start <= row_end && col_start <= col_end){
	        
	        for(int col=col_start; col<=col_end; col++){
	            System.out.print(matrix[row_start][col] + " ");
	        }
	        
	        row_start ++;
	        
	        for(int row=row_start; row<=row_end; row++){
	            System.out.print(matrix[row][col_end] + " ");
	        }
	        
	        col_end --;
	        
	        for(int col=col_end; col>=col_start; col--){
	            System.out.print(matrix[row_end][col] + " ");
	        }
	        
	        row_end --;
	        
	        for(int row=row_end; row>=row_start; row--){
	            System.out.print(matrix[row][col_start] + " ");
	        }
	        
	        col_start ++;
	        
	        System.out.println();
	   }
	}
}
